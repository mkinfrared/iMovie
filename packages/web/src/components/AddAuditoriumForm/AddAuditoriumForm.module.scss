@import "src/styles/common";
@import "src/styles/colors";

.AddAuditoriumForm {

  .container {
    display: flex;

    form,
    .preview {
      padding: 20px;
      width: 50%;
    }

    .inputField {
      @extend .input;
      display: flex;
      flex-wrap: nowrap;
      justify-content: space-between;

      &:not(:last-child) {

        .addRemoveButton {
          visibility: hidden;
        }
      }

      > div {
        width: 75%;
      }

      .addRemoveButton {
        height: 55px;
        width: 55px;
      }

      .select {
      }
    }
  }
}
