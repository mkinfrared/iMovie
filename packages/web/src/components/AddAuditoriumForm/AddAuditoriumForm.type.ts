export interface AddAuditoriumFormProps {
  cinemaId: number;
}
