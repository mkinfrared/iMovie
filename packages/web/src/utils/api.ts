import axios, { AxiosError } from "axios";
import { OptionsObject, SnackbarKey, SnackbarMessage } from "notistack";

import NotificationAction from "ui/NotificationAction";
import { BASE_URL } from "utils/secrets";

type ShowSnackbar = (
  message: SnackbarMessage,
  options?: OptionsObject
) => SnackbarKey;

const api = axios.create({
  baseURL: BASE_URL,
  withCredentials: true
});

const configureInterceptors = (showSnackbar: ShowSnackbar) => {
  api.interceptors.response.use(
    (config) => config,
    (error: AxiosError) => {
      if (error.message === "Network Error") {
        showSnackbar("Network Error", {
          variant: "error",
          action: NotificationAction
        });

        return;
      }

      return Promise.reject(error);
    }
  );
};

export default api;

export { configureInterceptors };
