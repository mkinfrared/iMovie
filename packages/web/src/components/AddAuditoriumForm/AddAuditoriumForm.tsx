import { addAuditoriumValidation } from "@imovie/common";
import Button from "@material-ui/core/Button";
import IconButton from "@material-ui/core/IconButton";
import TextField from "@material-ui/core/TextField";
import Typography from "@material-ui/core/Typography";
import AddIcon from "@material-ui/icons/Add";
import ClearIcon from "@material-ui/icons/Clear";
import omit from "lodash/omit";
import React, { useCallback, useMemo, useState } from "react";
import { Controller, useForm } from "react-hook-form";

import Seats from "components/SeatsPreview";
import Select from "ui/Select";
import { Option } from "ui/Select/Select.type";
import api from "utils/api";
import getFormErrors from "utils/getFormErrors";

import css from "./AddAuditoriumForm.module.scss";
import { AddAuditoriumFormProps } from "./AddAuditoriumForm.type";

type FormData = Record<string, number> & {
  name: string;
};

const AddAuditoriumForm = ({ cinemaId, onCancel }: AddAuditoriumFormProps) => {
  const [rowsAmount, setRowsAmount] = useState(1);

  const {
    control,
    handleSubmit,
    register,
    errors,
    getValues,
    triggerValidation,
    setError,
    formState
  } = useForm<FormData>({
    validationSchema: addAuditoriumValidation,
    mode: "onChange"
  });

  const seatOptions = useMemo(() => {
    const result: Option[] = [];

    for (let i = 0; i < 100; i++) {
      result.push({ value: i, label: i.toString() });
    }

    return result;
  }, []);

  const onSubmit = handleSubmit(async ({ name, ...rows }) => {
    const data = { name, cinemaId, ...rows };

    try {
      await api.post("/auditorium", data);
    } catch (e) {
      if (e.response?.status === 409) {
        const responseErrors = getFormErrors(e.response?.data);

        setError(responseErrors);
      }
    }
  });

  const handleSelectChange = ([event]: any[]) => {
    triggerValidation();

    return event.target.value;
  };

  const handleAddRowClick = useCallback(
    (rowName: string) => async () => {
      const zCode = "Z".charCodeAt(0);
      const currentRowCode = rowName.charCodeAt(0);
      const addRowCode = currentRowCode + 1;

      if (addRowCode > zCode) {
        return;
      }

      const valid = await triggerValidation();

      if (valid) {
        setRowsAmount((prevState) => prevState + 1);
      }
    },
    [triggerValidation]
  );

  const handleRemoveRowClick = useCallback(
    (rowName: string) => () => {
      const aCode = "A".charCodeAt(0);
      const removeRowCode = rowName.charCodeAt(0);

      if (removeRowCode <= aCode) {
        return;
      }

      triggerValidation();

      setRowsAmount((prevState) => prevState - 1);
    },
    [triggerValidation]
  );

  const renderSelects = () => {
    const selects: React.ReactElement[] = [];

    for (let i = 0; i < rowsAmount; i++) {
      const aCode = "A".charCodeAt(0);
      const letter = String.fromCharCode(aCode + i);

      const element = (
        <div key={letter} className={css.inputField}>
          <div>
            <Controller
              control={control}
              className={css.select}
              name={letter}
              label={`Row ${letter}`}
              as={Select}
              options={seatOptions}
              onChange={handleSelectChange}
              error={!!errors[letter]}
              helperText={errors[letter]?.message}
            />
          </div>
          <IconButton
            className={css.addRemoveButton}
            onClick={handleAddRowClick(letter)}
            color="primary"
            data-testid="addRowButton"
          >
            <AddIcon />
          </IconButton>
          <IconButton
            className={css.addRemoveButton}
            onClick={handleRemoveRowClick(letter)}
            color="primary"
            data-testid="removeRowButton"
          >
            <ClearIcon />
          </IconButton>
        </div>
      );

      selects.push(element);
    }

    return selects;
  };

  const rows = omit(getValues(), "name");

  return (
    <div className={css.AddAuditoriumForm}>
      <Typography variant="h5">Add Auditorium</Typography>
      <div className={css.container}>
        <form onSubmit={onSubmit} className={css.form} data-testid="form">
          <TextField
            fullWidth
            name="name"
            className={css.input}
            label="Auditorium Name"
            variant="outlined"
            helperText={errors.name?.message || null}
            error={!!errors.name}
            inputRef={register}
            inputProps={{ "data-testid": "auditoriumName" }}
          />
          <div>{renderSelects()}</div>
          <Button onClick={onCancel}>Cancel</Button>
          <Button
            data-testid="submitButton"
            onClick={onSubmit}
            color="primary"
            disabled={formState.isSubmitting}
          >
            Add
          </Button>
        </form>
        <div className={css.preview}>
          <Seats rows={rows} />
        </div>
      </div>
    </div>
  );
};

export default React.memo(AddAuditoriumForm);
