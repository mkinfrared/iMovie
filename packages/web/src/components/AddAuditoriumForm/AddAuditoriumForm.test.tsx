import { act, fireEvent, render, within } from "@testing-library/react";
import React, { ReactElement } from "react";

import api from "utils/api";

import AddAuditoriumForm from "./AddAuditoriumForm";

jest.mock("utils/api");

describe("<AddAuditoriumForm />", () => {
  const apiMock = api as jest.Mocked<typeof api>;
  const cinemaId = 42;
  const onCancel = jest.fn();

  let Component: ReactElement;

  beforeEach(() => {
    Component = <AddAuditoriumForm cinemaId={cinemaId} onCancel={onCancel} />;
  });

  it("should be defined", () => {
    const { container } = render(Component);

    expect(container).toBeDefined();
  });

  it("should match a snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should submit the form", async () => {
    const { container, findByRole, getByTestId } = render(Component);

    const selectButton = container.parentNode?.querySelector(
      "[role=button]"
    ) as HTMLElement;

    const addButton = getByTestId("addRowButton");
    const removeButton = getByTestId("removeRowButton");
    const submitButton = getByTestId("submitButton");
    const auditoriumName = getByTestId("auditoriumName") as HTMLInputElement;

    fireEvent.change(auditoriumName, { target: { value: "marklar" } });

    expect(auditoriumName.value).toBe("marklar");

    fireEvent.click(addButton);

    let selects = container.querySelectorAll(".select");

    expect(selects).toHaveLength(1);

    fireEvent.click(removeButton);

    selects = container.querySelectorAll(".select");

    expect(selects).toHaveLength(1);

    fireEvent.mouseDown(selectButton);

    const listbox = await findByRole("listbox");
    const listItem = await within(listbox).findByText("7");

    fireEvent.click(listItem);

    await within(container).findByText("7");

    await act(async () => {
      await fireEvent.click(submitButton);
    });

    expect(apiMock.post).toHaveBeenCalled();
  });

  it("should set errors on form", async () => {
    const { container, findByRole, getByTestId, getByText } = render(Component);

    const selectButton = container.parentNode?.querySelector(
      "[role=button]"
    ) as HTMLElement;

    const addButton = getByTestId("addRowButton");
    const removeButton = getByTestId("removeRowButton");
    const submitButton = getByTestId("submitButton");
    const auditoriumName = getByTestId("auditoriumName") as HTMLInputElement;

    fireEvent.change(auditoriumName, { target: { value: "marklar" } });

    expect(auditoriumName.value).toBe("marklar");

    fireEvent.click(addButton);

    let selects = container.querySelectorAll(".select");

    expect(selects).toHaveLength(1);

    fireEvent.click(removeButton);

    selects = container.querySelectorAll(".select");

    expect(selects).toHaveLength(1);

    fireEvent.mouseDown(selectButton);

    const listbox = await findByRole("listbox");
    const listItem = await within(listbox).findByText("7");

    fireEvent.click(listItem);

    await within(container).findByText("7");

    const data = {
      name: ["they took'er jobs!!"]
    };

    apiMock.post.mockReturnValueOnce(
      Promise.reject({ response: { data, status: 409 } })
    );

    await act(async () => {
      await fireEvent.click(submitButton);
    });

    expect(apiMock.post).toHaveBeenCalled();

    expect(getByText(data.name[0])).toBeInTheDocument();
  });

  it("should not set errors on form", async () => {
    const { container, findByRole, getByTestId, queryByText } = render(
      Component
    );

    const selectButton = container.parentNode?.querySelector(
      "[role=button]"
    ) as HTMLElement;

    const addButton = getByTestId("addRowButton");
    const removeButton = getByTestId("removeRowButton");
    const submitButton = getByTestId("submitButton");
    const auditoriumName = getByTestId("auditoriumName") as HTMLInputElement;

    fireEvent.change(auditoriumName, { target: { value: "marklar" } });

    expect(auditoriumName.value).toBe("marklar");

    fireEvent.click(addButton);

    let selects = container.querySelectorAll(".select");

    expect(selects).toHaveLength(1);

    fireEvent.click(removeButton);

    selects = container.querySelectorAll(".select");

    expect(selects).toHaveLength(1);

    fireEvent.mouseDown(selectButton);

    const listbox = await findByRole("listbox");
    const listItem = await within(listbox).findByText("7");

    fireEvent.click(listItem);

    await within(container).findByText("7");

    const data = {
      name: ["they took'er jobs!!"]
    };

    apiMock.post.mockReturnValueOnce(
      Promise.reject({ response: { data, status: 500 } })
    );

    await act(async () => {
      await fireEvent.click(submitButton);
    });

    expect(apiMock.post).toHaveBeenCalled();

    const error = queryByText(data.name[0]);

    expect(error).toBeNull();
  });
});
