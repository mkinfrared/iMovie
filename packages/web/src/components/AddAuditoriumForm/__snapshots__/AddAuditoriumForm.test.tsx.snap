// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddAuditoriumForm /> should match a snapshot 1`] = `
<div>
  <div
    class="AddAuditoriumForm"
  >
    <h5
      class="MuiTypography-root MuiTypography-h5"
    >
      Add Auditorium
    </h5>
    <div
      class="container"
    >
      <form
        class="form"
        data-testid="form"
      >
        <div
          class="MuiFormControl-root MuiTextField-root input MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
          >
            Auditorium Name
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              data-testid="auditoriumName"
              name="name"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-7"
              >
                <span>
                  Auditorium Name
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div>
          <div
            class="inputField"
          >
            <div>
              <div
                class="MuiFormControl-root select MuiFormControl-fullWidth"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                  data-shrink="false"
                  id="select"
                >
                  Row A
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                >
                  <div
                    aria-haspopup="listbox"
                    aria-labelledby="select"
                    class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
                    role="button"
                    tabindex="0"
                  >
                    <span>
                      â€‹
                    </span>
                  </div>
                  <input
                    type="hidden"
                    value=""
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="PrivateNotchedOutline-legendLabelled-7"
                    >
                      <span>
                        Row A
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <button
              class="MuiButtonBase-root MuiIconButton-root addRemoveButton MuiIconButton-colorPrimary"
              data-testid="addRowButton"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiIconButton-root addRemoveButton MuiIconButton-colorPrimary"
              data-testid="removeRowButton"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Cancel
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
          data-testid="submitButton"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Add
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </form>
      <div
        class="preview"
      >
        <div
          class="Seats"
        >
          <svg
            class="screen"
            style="width: 0px;"
            viewBox="0 0 830 69.59"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M.5.5h829V69Q415,.5.5,69Z"
              stroke-miterlimit="10"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
